src/mumble/GlobalShortcut_win.cpp

```diff
@@ -130,576 +52,356 @@ GlobalShortcutWin::~GlobalShortcutWin() {
-		case WM_KEYUP:
-		case WM_SYSKEYDOWN:
-		case WM_SYSKEYUP:
-			break;
-		default:
-			return false;
+	auto buffer = std::make_unique< uint8_t[] >(size);
+	if (GetRawInputData(handle, RID_INPUT, buffer.get(), &size, sizeof(RAWINPUTHEADER)) <= 0) {
+		return;
 	}
 
-	DWORD scancode = (msg->lParam >> 16) & 0xff;
-	DWORD vkcode   = msg->wParam;
-	bool extended  = !!(msg->lParam & 0x01000000);
-	bool up        = !!(msg->lParam & 0x80000000);
-
-	InjectKeyboardMessageEvent *ikme = new InjectKeyboardMessageEvent(scancode, vkcode, extended, !up);
-	boost::future< bool > suppress   = ikme->shouldSuppress();
-	qApp->postEvent(this, ikme);
-	return suppress.get();
-}
-
-bool GlobalShortcutWin::injectMouseMessage(MSG *msg) {
-	if (!bHook) {
-		return false;
-	}
+	auto input = reinterpret_cast< const PRAWINPUT >(buffer.get());
+	switch (input->header.dwType) {
+		case RIM_TYPEMOUSE: {
+			const auto &mouse = input->data.mouse;
+			emit mouseMessage(mouse.usButtonFlags, mouse.usButtonData);
+			break;
+		}
+		case RIM_TYPEKEYBOARD: {
+			const auto &keyboard = input->data.keyboard;
+			if (keyboard.MakeCode == KEYBOARD_OVERRUN_MAKE_CODE) {
+				// Invalid or unrecognizable combination of keys is pressed or
+				// the number of keys pressed exceeds the limit for this keyboard.
+				return;
+			}
 
-	bool down        = false;
-	unsigned int btn = 0;
+			if (keyboard.VKey == 255) {
+				// Discard "fake keys" which are part of an escaped sequence.
+				return;
+			}
 
-	// Convert the Windows mouse message into a DirectInput
-	// button index, and store the pressed state of the button.
-	switch (msg->message) {
-		case WM_LBUTTONDOWN:
-			down = true;
-		case WM_LBUTTONUP:
-			btn = 3;
-			break;
-		case WM_RBUTTONDOWN:
-			down = true;
-		case WM_RBUTTONUP:
-			btn = 4;
+			emit keyboardMessage(keyboard.Flags, keyboard.MakeCode, keyboard.VKey);
 			break;
-		case WM_MBUTTONDOWN:
-			down = true;
-		case WM_MBUTTONUP:
-			btn = 5;
-			break;
-		case WM_XBUTTONDOWN:
-			down = true;
-		case WM_XBUTTONUP: {
-			unsigned int offset = (msg->wParam >> 16) & 0xffff;
-			btn                 = 5 + offset;
 		}
-		default:
-			// Non-mouse event. Return early.
-			return false;
-	}
+		case RIM_TYPEHID: {
+			const auto &hid = input->data.hid;
+			std::vector< char > reports(hid.dwSizeHid * hid.dwCount);
+			memcpy(&reports[0], hid.bRawData, reports.size());
 
-	InjectMouseMessageEvent *imme  = new InjectMouseMessageEvent(btn, down);
-	boost::future< bool > suppress = imme->shouldSuppress();
-	qApp->postEvent(this, imme);
-	return suppress.get();
+			emit hidMessage(input->header.hDevice, std::move(reports), hid.dwSizeHid);
+		}
+	}
 }
 
-bool GlobalShortcutWin::handleKeyboardMessage(DWORD scancode, DWORD vkcode, bool extended, bool down) {
-	GlobalShortcutWin *gsw = static_cast< GlobalShortcutWin * >(engine);
-
-	QList< QVariant > ql;
+void GlobalShortcutWin::on_hidMessage(const HANDLE deviceHandle, std::vector< char > reports,
+									  const uint32_t reportSize) {
```

Comment:
Why is the size of the vector explicitly passed in this function?

Reply:
It's the size of a single report. `reports.size()` returns the number of reports.

Reply:
That still seems odd. Shouldn't the vector be of type `vector< T >` where `T` is the type of a report? And in that case the size of a report should be `sizeof(T)` :thinking:

Reply:
`T` is `char`, because report structure is opaque.

Reply:
In that case I think we should either use `uint8_t` or even better: Typedef a type `byte` to be `uint8_t` and use that.

Reply:
I generally use `uint8_t`, but in this case it would require a `reinterpret_cast` when calling [`HidP_GetUsages()`](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/hidpi/nf-hidpi-hidp_getusages).

Reply:
Ah okay I see